### PR TITLE
Implement BigQuery dataset label manipulation

### PR DESCRIPTION
**Pull request by jskeet**
(Review commit-wise... first two commits are small.)

### CODE REVIEW

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryClientImpl.DatasetLabels.cs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+            GaxPreconditions.CheckNotNull(datasetReference, nameof(datasetReference));
+            GaxPreconditions.CheckNotNull(labels, nameof(labels));
+
+            int retries = options?.Retries ?? ModifyLabelsOptions.DefaultRetries;
+            var getOptions = new GetDatasetOptions { ETag = options?.ETag };
+
+            while (true)
+            {
+                // Outside the catch, as we don't need to retry if *this* fails the precondition.
+                // If the bucket originally has the "right" version but then changes before we patch it, we'll end up retrying this
```
Comment (evildour) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryClientImpl.DatasetLabels.cs]: "bucket" => "dataset"? There's a few of these below too

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/ModifyLabelsOptions.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+        /// <summary>
+        /// The number of times to retry the modification if the etag changes
+        /// in the read/modify/write cycle. If this property is not set, a suitable default is used.
+        /// The value must not be negative.
+        /// </summary>
+        /// <remarks>
+        /// Modifying labels involves reading the resource in one request,
+        /// then sending another request with the new labels, including an etag check
+        /// to ensure that the bucket hasn't changed between the two requests. If the etag
```
Comment (evildour) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/ModifyLabelsOptions.cs]: And one more here

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.GenerateOverloads/Methods/ModifyDatasetLabels.xml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    <Parameter Name="labels" Type="IDictionary&lt;string, string&gt;" Comment="The labels to set. Must contain at least one entry; keys must not be null, but values may be (indicating label removal)." />
+  </AdditionalParameters>
+
+  <Comments>
+    <summary>
+      Sets or removes one or more labels on {target}.
+    </summary>
+    <remarks>
+      Each entry in <paramref name="labels"/> is treated as a label to set or remove. If the value is null,
+      it will be removed from the bucket; otherwise, it will be set/added. Labels which do not have an entry
```
Comment (evildour) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.GenerateOverloads/Methods/ModifyDatasetLabels.xml]: Oops, meant the reverse of what I said

Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.GenerateOverloads/Methods/ModifyDatasetLabels.xml]: Whoops - yes indeed. At which point, the beauty of the system is that I hardly need to change anything :)

**Comment by jskeet** [comment]
All buckets fixed - thanks Mike :)

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.GenerateOverloads/Methods/ClearDatasetLabels.xml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+﻿<Method Name="ClearDatasetLabels"
+        TargetType="Dataset"
+        ReturnType="IDictionary&lt;string, string&gt;">
```
Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.GenerateOverloads/Methods/ClearDatasetLabels.xml]: Only just seen this one. We can change that later if we want... it's *relatively* rare, and this is only for BigQuery. Going for cheap and cheerful where possible :)

**Pull request merged**